pkg/roachprod/vm/gce/gcloud.go

```diff
@@ -473,6 +481,225 @@ func (p *Provider) GetVMSpecs(
 	return vmSpecs, nil
 }
 
+func (p *Provider) GetVMArchitecture(l *logger.Logger, v *vm.VM) (vm.CPUArch, error) {
+
+	// If VM is tagged with arch FIPS, we use it
+	archLabel, ok := v.Labels["arch"]
+	if ok && archLabel == "fips" {
+		return vm.ArchFIPS, nil
+	}
+
+	// Otherwise, we query the CPU architecture from the machine type
+	return p.getCPUArchFromMachineType(v.MachineType, v.Zone, WideCacheKey)
+}
+
+func (p *Provider) getCPUArchFromMachineType(
+	machineType, zone string, ckl CacheKeyLookup,
+) (vm.CPUArch, error) {
+
+	// Note: zone is specified as an empty string as machine type architectures are the same
+	// across all zones and we want to have as many cache hits as possible
+	metadata, err := p.getMachineTypeSpecs(machineType, p.GetProject(), zone, ckl)
+	if err != nil {
+		return vm.ArchUnknown, err
+	}
+
+	// If arch is unspecified, GCP's default is amd64
+	if metadata.Architecture == "" {
+		return vm.ParseArch(string(vm.ArchAMD64)), nil
+	}
+
+	// Sanitize returned arch and return
+	return vm.ParseArch(metadata.Architecture), nil
+}
+
+func (p *Provider) getMachineTypeSpecs(
+	machineType, project, zone string, ckl CacheKeyLookup,
+) (*MachineTypeSpecs, error) {
+	return p.machineTypeSpecsCache.fetchFromProviderWithCache(machineType, project, zone, ckl)
+}
+
+type ProviderMachineTypeSpecsCache struct {
+	syncutil.Mutex
+	cache map[string]*MachineTypeSpecs
+}
+
+type MachineTypeSpecs struct {
+	Architecture                 string
+	Description                  string
+	GuestCpus                    int
+	Kind                         string
+	MaximumPersistentDisks       int
+	MaximumPersistentDisksSizeGb string
+	MemoryMb                     int
+	Name                         string
+	Zone                         string
+}
+
+type CacheKeyLookup string
+
+const (
+	NarrowCacheKey CacheKeyLookup = "NARROW_CACHE_KEY"
+	WideCacheKey   CacheKeyLookup = "WIDE_CACHE_KEY"
+)
+
+// fetchFromProviderWithCache fetches the machine-type's specs from the provider
+// and caches the result to return it on later calls.
+// As the machine type availability depends on the zone, the caching of each call
+// happens on both a `WideCacheKey` and a `NarrowCacheKey` cache key. This allows
```

Comment:
On a first pass, I confess I'm not a huge fan of `WideCacheKey` and `NarrowCacheKey`; I simply find them confusing. Part of the issue is that there's a lot going on in this function. Having a separate function per type of lookup (e.g., `fetchXXX` vs. `fetchXXXWithZone`) would make it simpler and less error-prone, imo. 
Let me sleep on this until the second pass.

Reply:
Done with the second pass. I'll take a fresh look tomorrow. As it stands, I still think this is error-prone. A consistent caching implementation shouldn't affect correctness. Yet, if we invoke this from `Create` and pass `WideCacheKey` (by mistake), there is a chance that we could hit the cache, when in fact `machineType` is unavailable in `zone`.